strawberryfields/gbs/sample.py

```diff
@@ -332,37 +372,70 @@ def gaussian(
     if not 0 <= loss <= 1:
         raise ValueError("Loss parameter must take a value between zero and one")
 
-    n_modes = len(alpha)
+    n_modes = len(t)
+
+    # initialize engine
 
     eng = sf.LocalEngine(backend="gaussian")
-    gbs = sf.Program(n_modes)
 
+    # initialize program objects
+
+    if np.all(t != 0):
+        gbs = sf.Program(n_modes * 2)
+    else:
+        gbs = sf.Program(n_modes)
+
+    # start circuit
     # pylint: disable=expression-not-assigned,pointless-statement
     with gbs.context as q:
 
-        # this interferometer has no action, but is kept to follow the decomposition given in the
-        # docstring
-        sf.ops.Interferometer(U1) | q
+        # two-mode squeezing gates
+
+        if np.all(t != 0):
+            for i in range(n_modes):
+                sf.ops.S2gate(t[i]) | (q[i], q[i + n_modes])
+
+        # first interferometer
+
+        sf.ops.Interferometer(U1) | q[:n_modes]
+
+        # squeezing gates
 
         for i in range(n_modes):
             sf.ops.Sgate(r[i]) | q[i]
 
-        sf.ops.Interferometer(U2) | q
+        # second interferometer
+
+        sf.ops.Interferometer(U2) | q[:n_modes]
+
+        # displacement gates
 
         for i in range(n_modes):
             sf.ops.Dgate(alpha[i]) | q[i]
 
+        # apply loss
+
         if loss:
             for _q in q:
                 sf.ops.LossChannel(1 - loss) | _q
```

Comment:
Does it still make sense to have loss now that we have made this function quite specific to vibronic spectra? I.e., would we realise a device in this way - two mode squeezing and operations on the first N modes.

@nquesada (I think?) has a way to map to 2N modes consisting of squeezing on all modes, a 2N mode interferometer, + displacement on all modes. Would this be the device we'd actually look to make - where in that case it might make more sense to talk about sampling but for now maybe not

Reply:
You got it all correct @trbromley ! Once it is mapped to the actual device it makes complete sense to put loss